### PR TITLE
Fixed GIT_DELTA_CONFLICTED not returned in some cases

### PR DESCRIPTION
**Pull request by swisspol**
(see commit message)

### CODE REVIEW

**Comment by swisspol** [comment]
This is the fallout from #3139, see the discussion there. I've reviewed carefully all the diff code paths, and I think that was the only occurence of this bug.

**Comment by carlosmn** [comment]
It would really help to have a test for this so it doesn't go unnoticed if it breaks again.

**Comment by swisspol** [comment]
All right, rebased PR, updated message and added unit test!

File: tests/diff/index.c [hunk 1]
```diff
@@ -239,3 +239,34 @@ void test_diff_index__reports_conflicts_when_reversed(void)
...
+
+	cl_git_pass(git_repository_index(&index, g_repo));
+	cl_git_pass(git_index_read_tree(index, a));
+
+	ancestor.path = ours.path = theirs.path = "file_not_in_head";
+	ancestor.mode = ours.mode = theirs.mode = 0100644;
+	git_oid_fromstr(&ancestor.id, "deadbeefdeadbeefdeadbeefdeadbeefdeadbeef");
+	git_oid_fromstr(&ours.id, "deadbeefdeadbeefdeadbeefdeadbeefdeadbeef");
+	git_oid_fromstr(&theirs.id, "deadbeefdeadbeefdeadbeefdeadbeefdeadbeef");
+	cl_git_pass(git_index_conflict_add(index, &ancestor, &ours, &theirs));
```
Comment (carlosmn) [inline tests/diff/index.c]: If the file does not exist in HEAD, then there won't be an 'ours' entry in the conflict, and possibly not one in 'ancestor' either.

File: tests/diff/index.c [hunk 2]
```diff
@@ -239,3 +239,34 @@ void test_diff_index__reports_conflicts_when_reversed(void)
...
+	git_diff *diff;
+	const git_diff_delta *delta;
+
+	git_tree *a = resolve_commit_oid_to_tree(g_repo, a_commit);
+
+	cl_git_pass(git_repository_index(&index, g_repo));
+	cl_git_pass(git_index_read_tree(index, a));
+
+	ancestor.path = ours.path = theirs.path = "file_not_in_head";
+	ancestor.mode = ours.mode = theirs.mode = 0100644;
```
Comment (carlosmn) [inline tests/diff/index.c]: We have `GIT_FILEMODE_BLOB` so we don't sprinkle the code with numbers.

**Comment by swisspol** [comment]
Rebased and fixed.

**Comment by ethomson** [comment]
Nice.  Thanks for the unit test here, this is a nice fix.

**Pull request merged**